File: backend/golang/blockchain/block.go
package blockchain

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/unicom-jf/blockchain/backend/golang/utils"
)
type Block struct {
	Timestamp int64 `json:"timestamp"`
	LastHash string `json:"lastHash"`
	Hash string `json:"hash"`
	//Data string `json:"data"`
	Data [][]byte `json:"data"`
	Difficulty int `json:"difficulty"`
	Nonce int `json:"nonce"`
}


func GenesisBlock() Block {
	return Block{
		Timestamp: 0, //milli seconds
		LastHash: "genesis_last_hash",
		Hash: "genesis_hash",
		Data: [][]byte{},
		Difficulty: 3,
		Nonce: 0,
	}
}

// compare two milliseconds
func AdjustDifficulty(last_block Block, timestamp int64) int {
	//used to many time
	if timestamp - last_block.Timestamp < utils.MineRate {
		return last_block.Difficulty + 1
	}
	if last_block.Difficulty == 1 {
		return 1
	}
	return last_block.Difficulty - 1
}
func Mine(last_block Block, data [][]byte) (Block, error) {
	last_hash := last_block.Hash
	nonce := last_block.Nonce
	//fmt.Printf("last_timestamp: %v\n", last_block.Timestamp)
	buf, err := json.Marshal(data)
	if err != nil {
		return Block{}, err
	}
	for {
		timestamp := time.Now().UnixMilli()
		difficulty := AdjustDifficulty(last_block, timestamp)
		hash, err := utils.Crypto_hash(strconv.FormatInt(timestamp, 10), 
			last_hash, string(buf), strconv.Itoa(difficulty), strconv.Itoa(nonce))
		if err != nil {
			log.Fatal(err)
		}
		if utils.Hex_to_binary(hash)[0:difficulty] == strings.Repeat("0", difficulty) {
			//fmt.Printf("timestamp: %v\n", timestamp)
			return Block {
				Timestamp: timestamp,
				LastHash: last_hash,
				Hash: hash,
				Data: data,
				Difficulty: difficulty,
				Nonce: nonce,
			}, nil
		}
		nonce++
	}
}

func BlockIsValid(last_block Block, block Block) error {
	if block.LastHash != last_block.Hash {
		return fmt.Errorf("the block's last_hash must be correct")	
	}
	if utils.Hex_to_binary(block.Hash)[0:block.Difficulty] != strings.Repeat("0", block.Difficulty) {
		return fmt.Errorf("the POW requirement is not met")	
	}
	if math.Abs(float64(block.Difficulty - last_block.Difficulty)) > 1 {
		return fmt.Errorf("the block's difficulty must only adjust by 1")	
	}
	buf, err := json.Marshal(block.Data)
	if err != nil {
		return err
	}
	hash, err := utils.Crypto_hash(strconv.FormatInt(block.Timestamp, 10), 
		block.LastHash, string(buf), strconv.Itoa(block.Difficulty), strconv.Itoa(block.Nonce))
	if err != nil {
		return err
	}
	if block.Hash != hash {
		return fmt.Errorf("the block's hash must be correct")	
	}
	
	return nil
}

File: backend/nodejs/blockchain/block.ts
/*
const GENESIS_DATA = {

}
*/
import { crypto_hash } from "../utils/crypto_hash";
import { hex_to_binary } from "../utils/hex_to_binary";
import { MINE_RATE } from "../config";
const GENESIS_BLOCK: Block = {
  timestamp: 0,
  lastHash: "last_hash",
  hash: "hash",
  data: [],
  difficulty: 3,
  nonce: 0,
};
class Block {
  timestamp!: number;
  lastHash: string;
  hash!: string;
  data!: any[];
  difficulty!: number;
  nonce!: number;

  constructor(
    timestamp: number,
    last_hash: string,
    hash: string,
    data: any[],
    difficulty: number,
    nonce: number
  ) {
    this.timestamp = timestamp;
    this.lastHash = last_hash;
    this.hash = hash;
    this.data = data;
    this.difficulty = difficulty;
    this.nonce = nonce;
  }

  static getGenesisBlock() {
    const { timestamp, lastHash, hash, data, difficulty, nonce } =
      GENESIS_BLOCK;
    return new Block(timestamp, lastHash, hash, data, difficulty, nonce);
    //return new Block(0, "last_hash", "hash", [], 3, 0);
    //return GENESIS_BLOCK;
  }

  static adjustDifficulty(last_block: Block, timestamp: number) {
    if (timestamp - last_block.timestamp < MINE_RATE) {
      return last_block.difficulty + 1;
    }
    if (last_block.difficulty == 1) {
      return last_block.difficulty;
    }
    return last_block.difficulty - 1;
  }
  static mine(last_block: Block, data: any[]) {
    let nonce = 0;
    let last_hash = last_block.hash;
    while (true) {
      let timestamp = Date.now();
      let difficulty = Block.adjustDifficulty(last_block, timestamp);
      let hash = crypto_hash(timestamp, last_hash, data, difficulty, nonce);
      if (
        hex_to_binary(hash).substring(0, difficulty) ===
        Array(difficulty + 1).join("0")
      ) {
        return new Block(timestamp, last_hash, hash, data, difficulty, nonce);
      }
      nonce++;
    }
  }
  static blockIsValid(last_block: Block, block: Block) {
    if (block.lastHash !== last_block.hash) {
      throw new Error("The block's last_hash must be correct");
    }
    if (
      hex_to_binary(block.hash).substring(0, block.difficulty) !==
      Array(block.difficulty + 1).join("0")
    ) {
      throw new Error("The POW requirement is not met");
    }
    if (Math.abs(block.difficulty - last_block.difficulty) > 1) {
      throw new Error("The block's difficulty must only adjust by 1");
    }
    if (
      block.hash !=
      crypto_hash(
        block.timestamp,
        block.lastHash,
        block.data,
        block.difficulty,
        block.nonce
      )
    ) {
      throw new Error("The block's hash must be correct");
    }
  }
}
export { Block, GENESIS_BLOCK };
